CMake improvements

xsacha opened the pull request:
**General Makefile->CMake convention changes**
CMake build type should not be set by the project. Fixed.
Build type automatic flags should be used instead of duplicated. Fixed.
Library type (static or shared) should not be set by the project. Fixed.
OpenMP flags should be detected automatically. Fixed.
Thread flags should be detected automatically. Fixed.
App should not be compiled (gen_config_h.exe) to generate a file. Fixed.

**Bug fixes**
If check failed on any non-Windows platform.
utest no longer runs (and fails) without fortran on non-MSVC.
GetArch no longer compiled and used if we are cross-compiling.
  :Now we set defines as getarch would do, based on target_core (this should be default in future)
Improve thread support and detection within CMake.
Added kernel/Makefile.LA to the CMake (TODO: Fails on DYNAMIC_ARCH).

**Feature Improvements**
Can now technically build any arch and any platform as host
Can now cross-compile (only added support for ARMv7 and ARMv8 so far but it should be as easy as dropping in the config).
Android builds successfully for ARMv7/ARMv8.

**CI Improvements**
Added CMake to Linux CI. Builds with GCC and Clang.
Also added a build without fortran.

**Known issues (also existed prior to this PR)**
Linux fails on DYNAMIC_ARCH due to kernel/Makefile.LA (use of undeclared identifier 'sneg_tcopy_EXCAVATOR')
Cross compile only works for the targets I added (ARMv7 and ARMv8).

Commit:
Let CMake deal with build type.

Commit:
Allow CMake to determine if it is building static or shared.

Commit:
Drop some redundant vars and improve arch detection in CMake.

Commit:
Fix bug that required fortran. Fix bug that needed CXX var. Remove redundant set vars. Fix threading detection. Do not attempt to run code if cross compiling.

martin-frbg commented:
There is pthreads as well for what it's worth, so I do not think the OpenMP part of your PR is correct, and I am not sure "allow cmake to determine if it is building static or shared" does quite the same as the current code either.
You can look at the options in Makefile.rule if you are not yet familiar with OpenBLAS.

xsacha committed:
Fix typos and use CMake OpenMP support.

xsacha commented:
FindThreads currently finds pthreads and works correctly on the systems I tried it on and most likely works on all systems correctly as it uses CMakes inbuilt checks.

CMake is designed to work this way. You can still create both static and shared in the same way it was being done before but they are separate configurations. It only needs to run the linker again to build it as static or shared depending on BUILD_SHARED_LIBS.

xsacha commented:
The only platform this CMake is currently working on as of 0.2.20 was Windows. On any *nix system it was failing in simple cmake if checks, let alone failing if Fortran was disabled or threading was enabled.

Still a while to go before it is usable on anything but Windows but much closer than it was.

xsacha committed:
Fix threading usage in CMake: s/SMP/USE_THREAD/

xsacha committed:
Only run utest without NOFORTRAN, same as Makefile. Linux now compiles.

xsacha committed:
Remove _static usages for tests

xsacha committed:
Add a CMake GCC and Clang target to Travis CI

xsacha committed:
Add kernel/Makefile.LA to CMake

xsacha committed:
Add support for cross compiling.
Add support for not having host compiler as CMake cannot detect such a compiler.
Add support for not using getarch.
Successfully builds Android ARMV8. Any target can be added by supplying the TARGET_CORE config in prebuild.cmake.

xsacha committed:
Improvements to previous commit (cross-compile).
Fix typos and bad if statements discovered in 0.2.20.

xsacha committed:
Clean up, fix old typos. Simplify arch usages. Move system arch check to earlier position.

iaincoulter commented:
Hi I copied these 18 changes and applied them to 0.2.20 release.  One issue i noticed in toplevel CMakeLists.txt is line 199.  ${CMAKE_CURRENT_SOURCE_DIR}/config.h  this file does not exist. When doing a CMake build the actual generated file is in ${CMAKE_BINARY_DIR}/config.h  (this also has a knock on effect in that any file including "common.h" reports "fatal error C1083: Cannot open include file: 'config.h': No such file or directory" (but this should be simple enough to add include directory pointing to CMAKE_BINARY_DIR.

PS Glad awk/sed gone but notice windows build still expects perl to be installed

xsacha commented:
Not sure why you don't see config.h as that was unchanged. Also in cross-compile I purposely move the written config to config.h. I think the successful Windows and Linux builds in CI are evidence that it works fine. Which configuration did not generate a config?

Also, I don't have perl -- what is this used for?

martin-frbg commented:
exports/gensymbol (used for preparing the openblas.def file on Windows) see end  of cmake/exports.cmake. The other users (f_check, c_check) have already been ported to (some level of) cmake magic some time ago

xsacha commented:
Can't you just use CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE
to autogenerate the def file?

iaincoulter commented:
Does CI environment clean all files between runs?  If not could an old config.h exist from before your cmake changes.   Looking at https://github.com/xianyi/OpenBLAS  there is no config.h.

Are you running cmake like this    'cmake -G "Visual Studio 14 Win64" . ' ? (the dot at the end is the important bit!)
Most other cmake projects recommend something similar like the following:

mkdir build
cd OpenBLAS
cmake -G "Visual Studio 14 Win64"  ../build

With this the original OpenBLAS folder should have zero changes applied to it.  All build/compile/etc will occur in the 'build' folder.

Adding "include_directories(${CMAKE_BINARY_DIR})" to CMakeLists.txt resolves the issue of not finding config.h

xsacha commented:
The CI always cleans between runs and builds in build directory. It also finds config.h, the same as all my local installs.

By the way, none of this being discussed has been changed whatsoever by this PR. I haven't touched Windows CI (appveyor), build dir or config.h location.

Edit: Also the build directory can be whatever you want. For example: PS C:\work\OpenBLAS\build\win> cmake ../.. -G"Visual Studio 15 2017 Win64" && cmake --build . --config Release
also works fine.

This PR has been tested on MSVC (cl), MSVC (clang-cl), mingw64, Linux (gcc), Linux (clang), Android (armv7/arm64) (clang), Android (armv7/arm64) (gcc) with clean builds.

xsacha commented:
By the way, none of this being discussed has been changed whatsoever by this PR.

iaincoulter commented:
xsacha I think the difference stems from fact I used 0.2.20 release and not develop branch. There are other cmake changes that may be cause for the issues I saw.  
From a pure cmake clean build folder way of doing things you would not typically modify anything in the source files location.  Eg for a completely clean cmake build, the contents of OpenBLAS folder would be identical before and after the build, but I don't know if thats a definitive rule when using cmake.

xsacha commented:
Yes you should never build a CMake project with source dir == binary dir.

I can add a fix so that Perl is not required for MSVC, however, this will be conditional on a newer CMake version and so the old functionality will remain if CMake < 3.4.

xsacha committed:
Do not require Perl for MSVC if CMake >= 3.4

isuruf commented:
https://github.com/xianyi/OpenBLAS/pull/1259 added support for out-of-source builds and a few other things, so you have to use that PR too.

isuruf commented:
.def file generated using perl added aliases for the functions. `dgemm = dgemm_`, `DGEMM = dgemm_`, etc. This would change that, although I don't know if these aliases are needed or not.

isuruf commented on CMakeLists.txt:
```diff
@@ -188,50 +166,33 @@ foreach (OUTPUTCONFIG ${CMAKE_CONFIGURATION_TYPES})
+  if(NOT NO_CBLAS)
+    add_subdirectory(ctest)
+  endif()
+elseif(MSVC)
```
Does this mean no `utest` on non-windows platforms when fortran is disabled?

xsacha replied:
@isuruf That's correct. utest was failing when fortran was disabled. This change allows build to succeed.

Is it meant to work without fortran because I don't think it ever has (in CMake).

Edit: On Windows, utest compiles but running it provides:
.\utest\openblas_utest.exe
RESULTS: 0 tests (0 ok, 0 failed, 0 skipped)

xsacha commented:
This PR is already based on that PR.
You don't want source directory polluted by build so config.h should not be inside source dir.

Edit: No, you're right. CMake won't generate the aliases. That was from importing the other .def.
Proper solution is to use declspec to export functions so no one has to worry about it.

xsacha committed:
Fix open_blas.config which was never working out-of-source. Remove need for gen_config_h.exe. If OpenMP is requested, do not silently ignore when it isn't available.

xsacha commented:
I have fixed your config.h issue, although it wasn't related to this PR, as it simplifies the CMake further (removes gen_config_h.exe) and allows it to work better with cross-compiling.

xsacha committed:
Clean up config file writing.

martin-frbg commented:
@isuruf any idea about the DYNAMIC_ARCH build failure ? Did it work for you back in august ? (I see the same failure even when I rollback to your commit, could be you PR back then was incomplete)

isuruf commented:
Hmm, which build failure? I see success on appveyor build.

martin-frbg commented:
When DYNAMiC_ARCH is set, the build aborts on setparam_EXCAVATOR.c with undefined identifiers like sneg_tcopy_EXCAVATOR

martin-frbg commented:
Seems the "if (NOT ${NO_LAPACK}" in kernel/CMakeLists.txt line 19 is not working as intended so common_lapack.h is not picked up for inclusion into kernel_TSUFFIX.h . Removing the curly braces seems to resolve this, however the DYNAMIC_ARCH build then (inexplicably?) stumbles over the lapack ?laswp_plus and ?laswp_minus, as the preprocessor gets invoked on common_params.h and
rejects all the arrow expressions in the xGEMM_UNROLL... defines as invalid preprocessor tokens.

martin-frbg commented:
So CMakefile is - and has always been - using the wrong file to generate the ?laswp for the dynamic builds (laswp_k.c which wants to use GEMM_UNROLL_N etc while the Makefile build uses laswp_k_4.c which has no such dependency). I will submit a PR for these "dynamic" issues later.
